tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -12,90 +12,457 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
+
 // Provide helper routine for obtaining  gpu target information useful
 // for llvm IR contruction.
 
 #include "tensorflow/compiler/xla/service/gpu/target_util.h"
 
+#include "absl/strings/str_cat.h"
 #include "llvm/IR/MDBuilder.h"
+#include "tensorflow/compiler/xla/primitive_util.h"
+#include "tensorflow/compiler/xla/service/llvm_ir/ir_builder_mixin.h"
 #include "tensorflow/compiler/xla/service/llvm_ir/llvm_util.h"
+#include "tensorflow/compiler/xla/xla_data.pb.h"
 #include "tensorflow/core/platform/logging.h"
 
 namespace xla {
+
+using absl::StrAppend;
+
 namespace gpu {
 namespace {
+
 // Utility functions to obtain NVPTX/AMDGPU specific information.
 
-// Wrapper structure for carrying llvm intrinsic ids for NVPTX/AMDGPU platforms.
-struct TargetIntrinsics {
-  llvm::Intrinsic::ID nvptx_intrinsic;
-  llvm::Intrinsic::ID amdgpu_intrinsic;
+// Wrapper structure for  intrinsic information for NVPTX/AMDGPU.
+struct TargetIntrinsicInfo {
+  TargetIntrinsicInfo(llvm::Intrinsic::ID x) : intrinsic(x) {}
+  llvm::Intrinsic::ID intrinsic;
+};
+
+// Wrapper structure for  device functionc information for NVPTX/AMDGPU.
+struct TargetFunctionInfo {
```

Comment:
This is the key struct in this change.  But I don't quite understand what you're trying to accomplish.  All of the options here need to be explained in more depth, then I can do a review of the logic.

- What is the reason for *not* checking the signature?
- What does PRIMITIVE_TYPE_INVALID mean in this context?
- Why are we using XLA primitive types instead of LLVM types?
etc.

Reply:
Broadly, the abstraction is trying to accommodate the differences between ptx/rocm. The differences can be 
       1. PTX calls an intrinsic while rocm needs a device function. This is the case for shfl. 
        2. PTX and rocm have different number of arguments. For shfl, PTX uses 4 arguments while rocm uses only 2. So target util provides the signature expected by rocm and how the call site operands map to the device function arguments. For the operands  from the call site that are not needed, target_util uses PRIMITIVE_TYPE_INVALID. 
       3. One another difference between PTX?/rocm is that rocm only provide i32 version of shfl. For other data types, conversions need to be generated. These are generated by going over the signature provided. 
        4. There are many cases, where ptx and rocm call in the same way such as math calls, various intrinsics. Since in these cases, there is no mis match expected, target util does not provide any signature to check.